### PR TITLE
KAFKA-12537: fix application shutdown corner case with only one thread

### PR DESCRIPTION
**Pull request by wcarlson5**
When in EOS the run loop terminates on that thread before the shutdown can be called. This is a problem for EOS single thread applications using the application shutdown feature. 

I changed it so in all cases with a single thread, the dying thread will spin up a new thread to communicate the shutdown and terminate the dying thread. Also @ableegoldman  redactored the catch blocks in runloop

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by mjsax — APPROVED** [review]
LGTM. Leave it to @ableegoldman to merge and cherry-pick.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Since we are adding a thread we won't have this issue anymore

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: If we add a thread before shutting down the application the thread with the exception can shutdown

**Review by wcarlson5 — COMMENTED** [review]
This remove the need to branching at all

**Comment by ableegoldman** [comment]
@wcarlson5 heads up, looks like you need to rebase this PR

**Comment by ableegoldman** [comment]
Also, since we have a fix for this, can we modify the `catch` blocks in the StreamThread loop to return false regardless of the processing mode? Also now that you explained how the default exception handler interacts with the StreamThread loop under ALOS, it seems like KAFKA-12537 probably affects both EOS and ALOS

**Comment by wcarlson5** [comment]
@ableegoldman fixed

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]:
```suggestion
                    log.warn("Attempt to shut down the application requires adding a thread to communicate the shutdown. No processing will be done on this thread");
```

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: Ah, nice that this solves the global thread issue as well! I guess technically this will still fail to communicate the shutdown if the application only ever runs the global thread and literally never started up any StreamThreads, but I think that's fine. Apparently running a global-only Streams app is a thing, as some users have reported in the past, but I would imagine this use case would almost certainly prefer the `REPLACE_THREAD` option.

~Ooh, wait. Do we need to add this check in the `REPLACE_THREAD` handling so we don't start up a StreamThread if it was the global thread that was killed?~ edit: we already do exactly this, but it's in `replaceThread()`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:629]: Since this thread is going to immediately shut down anyways, I think we can skip the `mainConsumer.enforceRebalance()`

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: WDYT about just checking for the case of a global-only topology before the `switch` statement, and just automatically invoking `closeToError()` with a warning that the other options are not supported in this case? We should also file an improvement ticket for the "restart the global thread" feature, if we don't already have one

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: We do check for the replace thread in the replace thread option. so we kinda already take care of it. If they do choose the shutdown the client we don't need to log a warning so its probably unnecessary. We don't have the option to replace the global threads yet. I will make a ticket for that.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:629]: This is also how we trigger the rebalance for the other threads so we can't just remove it

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: If all the Clients are global thread only shouldn't they hit the same error?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: Whoops, should've scrolled up a bit to the `replaceThread()` method. Thanks for filing the ticket

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: Well, who knows what the exception is -- could be some local disk error or corruption, or they're using a remote state store and god-knows-what happened. I agree that it's probably rare for a global-only app to hit an exception and fail to shut down the application, and not end up shutting down anyways due to hitting the same exception elsewhere. But anything's possible

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:629]: But do we even need to invoke `maybeSendShutdown` at all from the two catch blocks? The thread that we start up should handle this

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:502]: I'm not sure there's a good way to check if it's a global-only topology at the moment, so I'm with not putting in a check for that case. Was just a suggestion

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:629]: No we don't we can take those out

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: This PR LGTM but can you add a quick test for the case of a global-only topology with injected error + SHUTDOWN_APPLICATION handler? Just want to make sure nothing weird happens when you try to start up a StreamThread with no topology

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I wasn't sure about how to inject an error into an only global topology. But I figured what we really needed to test was if we can add a thread to a global-only topology. So I did that WDYT?

It will hit an illegal state exception as "Consumer is not subscribed to any topics or assigned any partitions" but that is after it will have resized the cache and after the shutdown rebalance has been triggered

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: You can use `StreamsBuilder#addGlobalStore` to inject an error, this method has a required parameter for a Processor which you're supposed to use to insert the data into the store (tbh no idea why you have to implement this yourself, it's caused a LOT of problems since it's unclear what you can/can't do with this -- but that's a whole other can of worms). 

Adding a thread clearly answered the question about what would happen, but imo we should just go ahead and use the `StreamsBuilder#addGlobalStore` + two KafkaStreams instances + injected error on one app to actually test the SHUTDOWN_APPLICATION feature.

Regarding the IllegalSTateException: wouldn't the thread hit this `IllegalStateException` before it got the chance to send the shutdown signal in a rebalance? IIRC the `subscribe` that's throwing is at the very beginning of the StreamThread loop. And unfortunately I think this may actually result in an endless cycle --  the new thread hits an unexpected exception and invokes the handler, SHUTDOWN_APPLICATION notices that this is the last StreamThread so it starts a new one, and so on.

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Well we know that the `SHUTDOWN_APPLICATION` will fail because the other client will not have a thread to receive the signal, unless you mean applications with different topologies?

It doesn't appear to endlessly loop the illegal state exception because both threads are still alive at that point but  it does not work as intended. I think that maybe this won't solve the global thread only case. We should only add a stream thread when there is one thread but not if there is zero. At least this ill take care of the eos problem and we can handle the global case as before

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Also the `IllegalStateException` is coming from the runLoop not the `subcribe`

**Comment by wcarlson5** [comment]
@ableegoldman I removed the global only case improvement and made https://issues.apache.org/jira/browse/KAFKA-12565.

Now this just has a fix for if there is one thread (eos or not) and consolidations of eos and alos logic divergence

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ah, I guess it's from trying to poll without any subscribed topics. Since that's not the case this alleviates my concern about the endless cycle too (In hindsight, yeah, no way `subscribe(<empty topics>)` would throw since that's supposed to be semantically identical to calling `unsubscribe()`)

**Review by ableegoldman — APPROVED** [review]
LGTM, will merge once the tests pass

**Comment by ableegoldman** [comment]
Looks like just one unrelated flaky test failure: `kafka.common.record.MemoryRecordsBuilderTest.[18] magic=1, bufferOffset=15, compressionType=GZIP`

**Comment by ableegoldman** [comment]
Merged to trunk, just running the tests before pushing to 2.8 @vvcephei 

edit: done

**Pull request merged**
